src/main/java/org/myrobotlab/framework/Service.java

```diff
@@ -1781,61 +1857,80 @@ public void setThisThread(Thread thisThread) {
     this.thisThread = thisThread;
   }
 
-  public ServiceInterface startPeer(String peerKey) {
-    String actualName = getPeerName(peerKey);
-    if (actualName == null) {
-      log.error("startPeer could not find actual name of {} in {}", peerKey, getName());
-    }
+  @Override
+  synchronized public ServiceInterface startPeer(String peerKey) {
+    // get current definition of config and peer
+    Peer peer = config.getPeer(peerKey);
 
-    ServiceInterface si = Runtime.start(actualName, null);
-    if (si != null) {
-      ServiceReservation sr = serviceType.getPeer(peerKey);
-      if (sr != null) {
-        sr.state = "STARTED";
-        sr.actualName = actualName;
-        sr.type = si.getSimpleName();
-        if (si != null) {
-          CodecUtils.setField(this, peerKey, si);
-        }
-      }
-      invoke("publishPeerStarted", peerKey);
-      broadcastState();
+    if (peer == null) {
+      error("startPeer could not find peerKey of %s in %s", peerKey, getName());
+      return null;
     }
-    return si;
-  }
-
-  public String getPeerType(String peerKey) {
-    ServiceReservation sr = serviceType.getPeer(peerKey);
-    return sr.type;
-  }
 
-  public void releasePeer(String peerKey) {
-    String actualName = getPeerName(peerKey);
-    Runtime.release(actualName);
-    ServiceReservation sr = serviceType.getPeer(peerKey);
-    if (sr != null) {
-      sr.actualName = actualName;
-      sr.state = "RELEASED";
-      invoke("publishPeerReleased", peerKey);
-    }
+    ServiceInterface si = Runtime.getService(peer.name);
+    if (si != null) {
+      // so this peer is already started, but are we responsible for
+      // all subpeers ?
+      return si;
+    }
+
+    // request to modify the plan's runtime to start all service that match
+    // actualName.*
+    Plan plan = Runtime.getPlan();
+    ServiceConfig sc = plan.get(peer.name);
+
+    if (sc == null) {
+      log.info("no current plan for peer {} - since this is a peer request we can make a plan", peer.name);
+      // error("plan.get(%s) == null", actualName);
+      Runtime.load(peer.name, peer.type);
+      sc = plan.get(peer.name);
+    }
+
+    // // recursive - start peers of peers of peers ...
+    // Map<String, Peer> subPeers = sc.getPeers();
+    // if (sc != null && subPeers != null) {
+    // for (String subPeerKey : subPeers.keySet()) {
+    // // IF AUTOSTART !!!
+    // Peer subPeer = subPeers.get(subPeerKey);
+    // if (subPeer.autoStart) {
+    // Runtime.start(sc.getPeerName(subPeerKey), subPeer.type);
+    // }
+    // }
+    // }
+
+    // start peer requested
+    Runtime.start(peer.name, sc.type);
     broadcastState();
+    return Runtime.getService(peer.name);
   }
 
-  @Override
-  @Deprecated /* use Runtime.start() */
-  public void loadAndStart() {
-    try {
-      load();
-      startService();
-    } catch (Exception e) {
-      log.error("Load and Start failed.", e);
+  synchronized public void releasePeer(String peerKey) {
```

Comment:
added javadocs ... the framework doesn't use releasePeer directly .. although it could ..
the recursive auto-releasing of peers happens here ..
https://github.com/MyRobotLab/myrobotlab/blob/config-with-peers-and-subscriptions/src/main/java/org/myrobotlab/service/Runtime.java#L2178-L2187  
default takes care of it all, a service would need to override to do something differently